modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -990,6 +1035,76 @@ def _union_all(
         HdkOnNativeDataframe
             The new frame.
         """
+        frames = [self] + other_modin_frames
+
+        # This is a special case, where we need to preserve the index of empty frames.
+        if any(len(f.columns) == 0 for f in frames):
+            return self._union_all_arrow(frames, join, sort, ignore_index)
+
+        # determine output columns
+        def join_cols():
+            new_cols_map = OrderedDict()
+            for col in self.columns:
+                new_cols_map[col] = self._dtypes[col]
+            for frame in other_modin_frames:
+                if join == "inner":
+                    for col in list(new_cols_map):
+                        if col not in frame.columns:
+                            del new_cols_map[col]
+                else:
+                    for col in frame.columns:
+                        if col not in new_cols_map:
+                            new_cols_map[col] = frame._dtypes[col]
```

Comment:
shouldn't this be something like this?
```suggestion
                        new_cols_map[col] = find_common_dtype([frame._dtypes[col], new_cols_map[col])
```

Reply:
The common type is selected below.